tests/integration/pilot/cni/cni_race_test.go

```diff
@@ -0,0 +1,158 @@
+// +build integ
+//  Copyright Istio Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package cni
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"path/filepath"
+	"testing"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	istioKube "istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/test/env"
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/cluster"
+	"istio.io/istio/pkg/test/framework/components/istio"
+	"istio.io/istio/pkg/test/framework/components/istioctl"
+	"istio.io/istio/pkg/test/framework/label"
+	"istio.io/istio/pkg/test/framework/resource"
+	"istio.io/istio/pkg/test/scopes"
+	"istio.io/istio/pkg/test/shell"
+	"istio.io/istio/pkg/test/util/retry"
+	"istio.io/istio/tests/integration/security/util"
+	"istio.io/istio/tools/istio-iptables/pkg/constants"
+)
+
+var (
+	ist  istio.Instance
+	apps = &util.EchoDeployments{}
+)
+
+func TestMain(m *testing.M) {
+	framework.NewSuite(m).
+		Label(label.CustomSetup).
+		Setup(istio.Setup(&ist, nil)).
+		Setup(func(t resource.Context) error {
+			return util.SetupApps(t, ist, apps, false)
+		}).
+		Run()
+}
+
+func TestCNIRaceRepair(t *testing.T) {
+	framework.NewTest(t).
+		Features("traffic.cni.race-condition-repair").
+		Run(func(t framework.TestContext) {
+			if !ist.Settings().EnableCNI {
+				t.Skip("CNI race condition mitigation is only tested when CNI is enabled.")
+			}
+			cluster := t.Clusters().Default()
+			// To begin with, delete CNI Daemonset to simulate a CNI race condition.
+			deleteCNIDaemonset(t, cluster)
+
+			// Rollout restart instances in namespace 1, and wait for a broken instance.
+			rolloutCmd := fmt.Sprintf("kubectl rollout restart deployment -n %s", apps.Namespace1.Name())
```

Comment:
There is a .Restart() on echo.Instance that may be preferred.

Reply:
Tricky thing is that workload will never become ready until CNI gets reinstalled. However `Restart` waits for workload to be ready, and I don't want to add a SkipWaiting to `Restart` just for this test. Let me know if you think it is fine to add skipWaiting option.